src/cmfd_solver.cpp

```diff
@@ -0,0 +1,361 @@
+        double res = (tmpx[irow] - x[irow]) / tmpx[irow];
+        err += res * res;
+      }
+    }
+
+    // Check convergence
+    err = std::sqrt(err / cmfd::dim);
+    if (err < tol)
+      return igs;
+
+    // Calculate new overrelaxation parameter
+    w = 1.0/(1.0 - 0.25 * cmfd::spectral * w);
+  }
+
+  // Throw error, as max iterations met
+  fatal_error("Maximum Gauss-Seidel iterations encountered.");
+
+  // Return -1 by default, although error thrown before reaching this point
+  return -1;
+}
+
+//==============================================================================
+// CMFD_LINSOLVER_NG solves a general CMFD linear system
+//==============================================================================
+
+int cmfd_linsolver_ng(const double* A_data, const double* b, double* x,
+                      double tol)
+{
+  // Set overrelaxation parameter
+  double w = 1.0;
+
+  // Perform Gauss-Seidel iterations
+  for (int igs = 1; igs <= 10000; igs++) {
+    double err = 0.0;
+
+    // Copy over x vector
+    std::vector<double> tmpx {x, x+cmfd::dim};
+
+    // Loop around matrix rows
+    for (int irow = 0; irow < cmfd::dim; irow++) {
+      // Get index of diagonal for current row
+      int didx = get_diagonal_index(irow);
+
+      // Perform temporary sums, first do left of diag, then right of diag
+      double tmp1 = 0.0;
+      for (int icol = cmfd::indptr[irow]; icol < didx; icol++)
+        tmp1 += A_data[icol] * x[cmfd::indices[icol]];
+      for (int icol = didx + 1; icol < cmfd::indptr[irow + 1]; icol++)
+        tmp1 += A_data[icol] * x[cmfd::indices[icol]];
+
+      // Solve for new x
+      double x1 = (b[irow] - tmp1) / A_data[didx];
+
+      // Perform overrelaxation
+      x[irow] = (1.0 - w) * x[irow] + w * x1;
+
+      // Compute residual and update error
+      double res = (tmpx[irow] - x[irow]) / tmpx[irow];
+      err += res * res;
+    }
+
+    // Check convergence
+    err = std::sqrt(err / cmfd::dim);
+    if (err < tol)
+      return igs;
+
+    // Calculate new overrelaxation parameter
+    w = 1.0/(1.0 - 0.25 * cmfd::spectral * w);
+  }
+
+  // Throw error, as max iterations met
+  fatal_error("Maximum Gauss-Seidel iterations encountered.");
+
+  // Return -1 by default, although error thrown before reaching this point
+  return -1;
+}
+
+//==============================================================================
+// OPENMC_INITIALIZE_LINSOLVER sets the fixed variables that are used for the
+// linear solver
+//==============================================================================
+
+extern "C"
+void openmc_initialize_linsolver(const int* indptr, int len_indptr,
+                                 const int* indices, int n_elements, int dim,
+                                 double spectral, const int* cmfd_indices,
+                                 const int* map)
+{
+  // Store elements of indptr
+  for (int i = 0; i < len_indptr; i++)
+    cmfd::indptr.push_back(indptr[i]);
```

Comment:
Should these be `clear()`'d first, in case user accidentally calls `openmc_initialize_linsolver` twice? Perhaps calling `free_memory_cmfd()` could make sense.